### PR TITLE
Add C detector samples.

### PR DESCRIPTION
**Pull request by saurabh-amazoncode-repo**
Add non compliant and compliant samples for C

### CODE REVIEW

File: c/src/detectors/c-os-command-injection/c-os-command-injection_compliant.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-os-command-injection@v1.0 defects=0}
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+int compliant(int argc, char** argv) {
+    //Compliant : Sanitize  and validate such inputs before passing.
```
Comment (bhaaayus) [inline c/src/detectors/c-os-command-injection/c-os-command-injection_compliant.c:9]: Where is the sanitization and validation being done?

File: c/src/detectors/c-os-command-injection/c-os-command-injection_non-compliant.c [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-os-command-injection@v1.0 defects=1}
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+int noncompliant(int argc, char **argv) {
+    //NonCompliant : Passing an untrusted input to a method that dynamically executes a command.
```
Comment (bhaaayus) [inline c/src/detectors/c-os-command-injection/c-os-command-injection_non-compliant.c:9]: The comment is misplaced

File: c/src/detectors/c-out-of-bounds-read/c-out-of-bounds-read_compliant.c [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-out-of-bounds-read@v1.0 defects=0}
+#include <stdio.h>
+
+int compliant() {
+    //Compliant : Perform bounds checking before accessing an array.
```
Comment (bhaaayus) [inline c/src/detectors/c-out-of-bounds-read/c-out-of-bounds-read_compliant.c:7]: The comment is misplaced

File: c/src/detectors/c-out-of-bounds-read/c-out-of-bounds-read_non-compliant.c [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-out-of-bounds-read@v1.0 defects=1}
+#include <stdio.h>
+
+int noncompliant() {
+    //NonCompliant : Code contains out-of-bounds read.
```
Comment (bhaaayus) [inline c/src/detectors/c-out-of-bounds-read/c-out-of-bounds-read_non-compliant.c:7]: Comment is misplaced

File: c/src/detectors/c-out-of-bounds-write/c-out-of-bounds-write_compliant.c [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-out-of-bounds-write@v1.0 defects=0}
+#include <stdio.h>
+
+int compliant(){
+    //Compliant : Perform bounds checking, validate input sizes and indices to ensure they stay within the allocated memory bounds.
```
Comment (bhaaayus) [inline c/src/detectors/c-out-of-bounds-write/c-out-of-bounds-write_compliant.c:7]: 1. The comment is misplaced
2. Please shorten the comment to a single line if possible

File: c/src/detectors/c-out-of-bounds-write/c-out-of-bounds-write_compliant.c [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  int id_sequence[4];
+
+  id_sequence[0] = 123;
+  id_sequence[1] = 234;
+  id_sequence[2] = 345;
+  id_sequence[3] = 456;
+
+  int n = (int) sizeof(id_sequence)/sizeof(id_sequence[0]);
```
Comment (bhaaayus) [inline c/src/detectors/c-out-of-bounds-write/c-out-of-bounds-write_compliant.c:15]: Add space before and after `/`

File: c/src/detectors/c-out-of-bounds-write/c-out-of-bounds-write_non-compliant.c [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-out-of-bounds-write@v1.0 defects=1}
+#include <stdio.h>
+
+void noncompliant(){
+    //NonCompliant : Code contains an out-of-bounds write which may lead to security issues.
```
Comment (bhaaayus) [inline c/src/detectors/c-out-of-bounds-write/c-out-of-bounds-write_non-compliant.c:7]: Comment is misplaced

File: c/src/detectors/c-path-traversal/c-path-traversal_compliant.c [hunk 8]
```diff
@@ -0,0 +1,35 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-path-traversal@v1.0 defects=0}
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+int compliant(int argc, char *argv[]) {
+    //Compliant : Validate user input against a predefined set of allowed directories or use secure input parsing techniques.
```
Comment (bhaaayus) [inline c/src/detectors/c-path-traversal/c-path-traversal_compliant.c:9]: 1. Comment is misplaced
2. Please shorten the comment to a single line possible

File: c/src/detectors/c-path-traversal/c-path-traversal_non-compliant.c [hunk 9]
```diff
@@ -0,0 +1,20 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-path-traversal@v1.0 defects=1}
+#include <stdio.h>
+
+void noncompliant() {
+    //NonCompliant : User input to specify file paths for reading or writing operation without proper validation.
```
Comment (bhaaayus) [inline c/src/detectors/c-path-traversal/c-path-traversal_non-compliant.c:7]: The comment is misplaced

File: c/src/detectors/c-random-fd-exhaustion/c-random-fd-exhaustion_compliant.c [hunk 10]
```diff
@@ -0,0 +1,26 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-random-fd-exhaustion@v1.0 defects=0}
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+int compliant() {
+    //Compliant : Manage and limit the resource allocations properly and implement appropriate error handling.
```
Comment (bhaaayus) [inline c/src/detectors/c-random-fd-exhaustion/c-random-fd-exhaustion_compliant.c:10]: The comment is misplaced

File: c/src/detectors/c-random-fd-exhaustion/c-random-fd-exhaustion_non-compliant.c [hunk 11]
```diff
@@ -0,0 +1,20 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-random-fd-exhaustion@v1.0 defects=1}
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+int noncompliant() {
+    //NonCompliant : Failure to limit and close open file descriptors allows uncontrolled resource consumption, potentially impacting system stability and performance.
```
Comment (bhaaayus) [inline c/src/detectors/c-random-fd-exhaustion/c-random-fd-exhaustion_non-compliant.c:10]: 1. Please shorten the comment to a single line
2. Comment seems to be misplaced

File: c/src/detectors/c-redundant-free-usage/c-redundant-free-usage_compliant.c [hunk 12]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-redundant-free-usage@v1.0 defects=0}
+#include <stdlib.h>
+
+int compliant() {
+    //Compliant : Avoid freeing the same memory address twice.
```
Comment (bhaaayus) [inline c/src/detectors/c-redundant-free-usage/c-redundant-free-usage_compliant.c:7]: Comment is misplaced

File: c/src/detectors/c-redundant-free-usage/c-redundant-free-usage_non-compliant.c [hunk 13]
```diff
@@ -0,0 +1,14 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-redundant-free-usage@v1.0 defects=1}
+#include <stdlib.h>
+
+int noncompliant() {
+    //NonCompliant : Calling `free(var)` twice on the same memory address, which may leads to program crashes.
```
Comment (bhaaayus) [inline c/src/detectors/c-redundant-free-usage/c-redundant-free-usage_non-compliant.c:7]: 1. Should be `... which may lead to program crashes.`
2. The comment seems to be misplaced

File: c/src/detectors/c-return-stack-address/c-return-stack-address_compliant.c [hunk 14]
```diff
@@ -0,0 +1,12 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-return-stack-address@v1.0 defects=0}
+#define STR_MAX 256
+int global_var = 42;
+
+int* compliant() {
+    //Compliant : Use heap-allocated memory, use static local variables when appropriate, or redesign the function to avoid returning local addresses.
```
Comment (bhaaayus) [inline c/src/detectors/c-return-stack-address/c-return-stack-address_compliant.c:8]: Please shorten the comment to a single line

File: c/src/detectors/c-return-stack-address/c-return-stack-address_non-compliant.c [hunk 15]
```diff
@@ -0,0 +1,9 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-return-stack-address@v1.0 defects=1}
+int* noncompliant() {
+    //NonCompliant : Insecure because it returns the address of a local variable, which becomes invalid after the function returns.
```
Comment (bhaaayus) [inline c/src/detectors/c-return-stack-address/c-return-stack-address_non-compliant.c:5]: Please shorten the comment to a single line

File: c/src/detectors/c-string-equality/c-string-equality_compliant.c [hunk 16]
```diff
@@ -0,0 +1,18 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-string-equality@v1.0 defects=0}
+#include <string.h>
+
+    char *s = "Hello";
```
Comment (bhaaayus) [inline c/src/detectors/c-string-equality/c-string-equality_compliant.c:6]: The indent looks misplaced

File: c/src/detectors/c-string-equality/c-string-equality_compliant.c [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+    char ss[] = "Hello";
+//Compliant : Compared the contents of two string variables using the `strcmp()`.
```
Comment (bhaaayus) [inline c/src/detectors/c-string-equality/c-string-equality_compliant.c:8]: The comment seems to be misplaced

File: c/src/detectors/c-string-equality/c-string-equality_non-compliant.c [hunk 18]
```diff
@@ -0,0 +1,18 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-string-equality@v1.0 defects=1}
+#include <string.h>
+#include <stdio.h>
+
+char *s = "Hello";
+char ss[] = "Hello";
+ //NonCompliant : Using `==` performs character comparison, Which could be leveraged to affect program security.
```
Comment (bhaaayus) [inline c/src/detectors/c-string-equality/c-string-equality_non-compliant.c:9]: In general, please add the comments on the line which is vulnerable, that would help users know what exactly is the problem

**Review by bhaaayus — CHANGES_REQUESTED** [review]

File: c/src/detectors/c-incorrect-use-sscanf-fn/c-incorrect-use-sscanf-fn_non-compliant.c [hunk 19]
```diff
@@ -0,0 +1,16 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+// {fact rule=c-incorrect-use-sscanf-fn@v1.0 defects=1}
+#include <stdio.h>
+#include <stdlib.h>
+
+int noncompliant() {
+    // Noncompliant : Avoid using `sscanf()` for number conversions.
```
Comment (bhaaayus) [inline c/src/detectors/c-incorrect-use-sscanf-fn/c-incorrect-use-sscanf-fn_non-compliant.c:9]: Please remove extra space after `Noncompliant`

File: c/src/detectors/c-insecure-use-scanf-fn/c-insecure-use-scanf-fn_non-compliant.c [hunk 20]
```diff
@@ -0,0 +1,12 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+// {fact rule=c-insecure-use-scanf-fn@v1.0 defects=1}
+#include <stdio.h>
+
+void noncompliant() {
+    char str[120];
+    // Noncompliant : Usage of `scanf()` can lead to 'buffer overflow' vulnerabilities.
```
Comment (bhaaayus) [inline c/src/detectors/c-insecure-use-scanf-fn/c-insecure-use-scanf-fn_non-compliant.c:9]: Please remove extra space after `Noncompliant`

File: c/src/detectors/c-insecure-temporary-file-or-directory/c-insecure-temporary-file-or-directory_compliant.c [hunk 21]
```diff
@@ -0,0 +1,21 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+// {fact rule=c-insecure-temporary-file-or-directory@v1.0 defects=0}
+#include <stdio.h>
+
+void compliant() {
+
+    // Compliant: Use secure function which manage the generation of unique and secure temporary files automatically.
```
Comment (bhaaayus) [inline c/src/detectors/c-insecure-temporary-file-or-directory/c-insecure-temporary-file-or-directory_compliant.c:9]: Can we shorten multiline comments to a single line? If not, can we maintain same indentation for better readability? Applicable to all files

File: c/src/detectors/c-insecure-use-scanf-fn/c-insecure-use-scanf-fn_compliant.c [hunk 22]
```diff
@@ -0,0 +1,18 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+// {fact rule=c-insecure-use-scanf-fn@v1.0 defects=0}
+#include <stdio.h>
+
+void compliant() {
+    char buf[8];
+
+    // Compliant: 'fgets' is a safer alternative
```
Comment (bhaaayus) [inline c/src/detectors/c-insecure-use-scanf-fn/c-insecure-use-scanf-fn_compliant.c:10]: API names should be enclosed in backticks

File: c/src/detectors/c-null-pointer-dereference/c-null-pointer-dereference_compliant.c [hunk 23]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-null-pointer-dereference@v1.0 defects=0}
+#include <stdio.h>
+
+void compliant()
+{
+   // Compliant: Pointers are always initialized before use.
+   int *ptr = NULL;
```
Comment (bhaaayus) [inline c/src/detectors/c-null-pointer-dereference/c-null-pointer-dereference_compliant.c:9]: The `if` condition will always be `false`. I think a better way would be to accept `*ptr` as an argument

File: c/src/detectors/c-insecure-temporary-file-or-directory/c-insecure-temporary-file-or-directory_compliant.c (hunk 21 above)
Comment (saurabh-amazoncode-repo) [reply c/src/detectors/c-insecure-temporary-file-or-directory/c-insecure-temporary-file-or-directory_compliant.c:9]: Hi,
Can you check this once. This is a single line comment.

File: c/src/detectors/c-incomplete-cleanup/c-incomplete-cleanup_compliant.c [hunk 24]
```diff
@@ -0,0 +1,26 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+// {fact rule=c-incomplete-cleanup@v1.0 defects=0}
+#include <stdio.h>
+#include <stdlib.h>
+
+int compliant(int print) {
+
+   // Compliant: Closing files and freeing resources when they are no longer needed.
```
Comment (bhaaayus) [inline c/src/detectors/c-incomplete-cleanup/c-incomplete-cleanup_compliant.c:10]: Do you think this should be present when the file is being closed on line 23. Please check on your side once.

File: c/src/detectors/c-insecure-temporary-file-or-directory/c-insecure-temporary-file-or-directory_non-compliant.c [hunk 25]
```diff
@@ -0,0 +1,16 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+// {fact rule=c-insecure-temporary-file-or-directory@v1.0 defects=1}
+#include <stdio.h>
+
+void noncompliant() {
+    // Noncompliant: Using insecure temporary files or directories leads to vulnerable attacks.
```
Comment (bhaaayus) [inline c/src/detectors/c-insecure-temporary-file-or-directory/c-insecure-temporary-file-or-directory_non-compliant.c:8]: Should be - `... leads to vulnerabilities or attacks.`

File: c/src/detectors/c-insecure-use-strcat-fn/c-insecure-use-strcat-fn_compliant.c [hunk 26]
```diff
@@ -0,0 +1,15 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+// {fact rule=c-insecure-use-strcat-fn@v1.0 defects=0}
+#include <string.h>
+#include <stdio.h>
+
+void compliant(char* src, char* dest, int dest_size) {
+    // Compliant: Applied bound checking or null terminator.
```
Comment (bhaaayus) [inline c/src/detectors/c-insecure-use-strcat-fn/c-insecure-use-strcat-fn_compliant.c:9]: There's an extra space before `//`

File: c/src/detectors/c-path-traversal/c-path-traversal_compliant.c [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+int compliant(int argc, char *argv[])
+{
+
+    char filename[100];
+
+    // Compliant: Validated user input.
```
Comment (bhaaayus) [inline c/src/detectors/c-path-traversal/c-path-traversal_compliant.c:14]: I think the validation happens on line 20

File: c/src/detectors/c-unsafe-file-extension/c-unsafe-file-extension_non-compliant.c [hunk 28]
```diff
@@ -0,0 +1,22 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+// {fact rule=c-unsafe-file-extension@v1.0 defects=1}
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+void noncompliant() {
+
+    //Noncompliant: Unsafe file extension vulnerabilities in your code can lead to significant security risks.
```
Comment (bhaaayus) [inline c/src/detectors/c-unsafe-file-extension/c-unsafe-file-extension_non-compliant.c:10]: Missing space after `//`

**Review by bhaaayus — CHANGES_REQUESTED** [review]

**Review by bhaaayus — APPROVED** [review]

**Review by atonaamz — APPROVED** [review]

**Pull request merged**